lib/Parse/ParseDecl.cpp

```diff
@@ -3236,15 +3236,36 @@ ParserResult<TypeDecl> Parser::parseDeclAssociatedType(Parser::ParseDeclOptions
     if (UnderlyingTy.isNull())
       return Status;
   }
-  
-  // Parse a 'where' clause if present. These are not supported, but we will
-  // get better QoI this way.
+
+  TrailingWhereClause *TrailingWhere = nullptr;
+  // Parse a 'where' clause if present.
   if (Tok.is(tok::kw_where)) {
-    GenericParamList *unused = nullptr;
-    auto whereStatus = parseFreestandingGenericWhereClause(
-        unused, WhereClauseKind::AssociatedType);
-    if (whereStatus.shouldStopParsing())
-      return whereStatus;
+    if (Context.isSwiftVersion3()) {
+      // Swift 3 doesn't support where clauses on associated types, so let's
+      // keep exactly the same parsing behaviour as it had.
+      GenericParamList *unused = nullptr;
+      auto whereStatus = parseFreestandingGenericWhereClause(
+          unused, WhereClauseKind::AssociatedType);
+      if (whereStatus.shouldStopParsing())
+        return whereStatus;
+    } else {
+      SourceLoc whereLoc;
+      SmallVector<RequirementRepr, 4> requirements;
+      bool firstTypeInComplete;
+      auto whereStatus =
+          parseGenericWhereClause(whereLoc, requirements, firstTypeInComplete);
+      if (whereStatus.isSuccess()) {
+        TrailingWhere =
+            TrailingWhereClause::create(Context, whereLoc, requirements);
+      } else if (whereStatus.hasCodeCompletion()) {
+        // FIXME: this is completely (hah) cargo culted.
+        if (CodeCompletion && firstTypeInComplete) {
+          CodeCompletion->completeGenericParams(nullptr);
+        } else {
+          return makeParserCodeCompletionResult<AssociatedTypeDecl>();
+        }
+      }
```

Comment:
Can we simplify the Swift 3 vs. Swift 4 logic here? Always go down the parsing path, but just drop `TrailingWhere` on the floor with a "this is not implemented yet; use -swift-version 4 to experiment" error of some sort?